scheduling/taskmetadata.go

```diff
@@ -0,0 +1,533 @@
+		if err != nil {
+			return TaskMetadataList{}, err
+		}
+		tml = append(tml, taskMetadata)
+	}
+	return tml, nil
+}
+
+func (m *TaskMetadataManager) CreateTaskRun(name string, taskID TaskID, trigger Trigger) (TaskRunMetadata, fferr.GRPCError) {
+	// ids will be generated by TM
+	taskRunKey := TaskRunKey{taskID: taskID}
+	taskMetadata, err := m.storage.Get(taskRunKey.String())
+	if err != nil {
+		err.AddDetail("task name", name)
+		return TaskRunMetadata{}, err
+	}
+
+	// Not sold on this naming for this struct. Maybe like RunHistory or something?
+	runs := TaskRuns{}
+	err = runs.Unmarshal([]byte(taskMetadata))
+	if err != nil {
+		err.AddDetail("task name", name)
+		err.AddDetail("task_id", string(taskID))
+		return TaskRunMetadata{}, err
+	}
+
+	latestID, err := runs.GetLatestRunId()
+	if err != nil {
+		err.AddDetail("task name", name)
+		err.AddDetail("task_id", string(taskID))
+		return TaskRunMetadata{}, err
+	}
+
+	startTime := time.Now().UTC()
+
+	metadata := TaskRunMetadata{
+		ID:          TaskRunID(latestID + 1),
+		TaskId:      taskID,
+		Name:        name,
+		Trigger:     trigger,
+		TriggerType: trigger.Type(),
+		Status:      Pending,
+		StartTime:   startTime,
+	}
+
+	runs.Runs = append(runs.Runs, TaskRunSimple{RunID: metadata.ID, DateCreated: startTime})
+
+	serializedRuns, err := runs.Marshal()
+	if err != nil {
+		err.AddDetail("task name", name)
+		err.AddDetail("task_id", string(taskID))
+		err.AddDetail("run_id", string(metadata.ID))
+		return TaskRunMetadata{}, err
+	}
+
+	serializedMetadata, err := metadata.Marshal()
+	if err != nil {
+		err.AddDetail("task name", name)
+		err.AddDetail("task_id", string(taskID))
+		err.AddDetail("run_id", string(metadata.ID))
+		return TaskRunMetadata{}, err
+	}
+	err = m.storage.Create(taskRunKey.String(), string(serializedRuns))
+	if err != nil {
+		err.AddDetail("task name", name)
+		err.AddDetail("task_id", string(taskID))
+		return TaskRunMetadata{}, err
+	}
+
+	taskRunMetaKey := TaskRunMetadataKey{taskID: taskID, runID: metadata.ID, date: startTime}
+	err = m.storage.Create(taskRunMetaKey.String(), string(serializedMetadata))
+	if err != nil {
+		err.AddDetail("task name", name)
+		err.AddDetail("task_id", string(taskID))
+		err.AddDetail("run_id", string(metadata.ID))
+		return TaskRunMetadata{}, err
+	}
+
+	return metadata, nil
+}
+
+func (m *TaskMetadataManager) GetRunByID(taskID TaskID, runID TaskRunID) (TaskRunMetadata, fferr.GRPCError) {
+	taskRunKey := TaskRunKey{taskID: taskID}
+	taskRunMetadata, err := m.storage.Get(taskRunKey.String())
+	if err != nil {
+		err.AddDetail("task_id", string(taskID))
+		err.AddDetail("run_id", string(runID))
+		return TaskRunMetadata{}, err
+	}
+
+	runs := TaskRuns{}
+	err = runs.Unmarshal([]byte(taskRunMetadata))
+	if err != nil {
+		err.AddDetail("task_id", string(taskID))
+		err.AddDetail("run_id", string(runID))
+		return TaskRunMetadata{}, err
+	}
+
+	// Want to move this logic out
+	found, runRecord := runs.ContainsRun(runID)
```

Comment:
hmmm i don't think this is going to be inefficient yet it could def be optimized.I think TaskRuns.runs should be a map of runId to Runs so it's easily searched

Reply:
actually now that i'm looking at it more it should definitely be that